src/NLog/Config/LoggingConfigurationParser.cs

```diff
@@ -1295,9 +1278,122 @@ private static string GetName(Target target)
             return string.IsNullOrEmpty(target.Name) ? target.GetType().Name : target.Name;
         }
 
+        private class ValidatedConfigurationElement : ILoggingConfigurationElement
+        {
+            private static readonly IDictionary<string, string> EmptyDefaultDictionary = new SortHelpers.ReadOnlySingleBucketDictionary<string, string>();
+
+            private readonly ILoggingConfigurationElement _element;
+            private readonly bool _throwConfigExceptions;
+            private IList<ValidatedConfigurationElement> _validChildren;
+
+            public static ValidatedConfigurationElement Create(ILoggingConfigurationElement element, LogFactory logFactory)
+            {
+                if (element is ValidatedConfigurationElement validConfig)
+                    return validConfig;
+                else
+                    return new ValidatedConfigurationElement(element, logFactory.ThrowConfigExceptions ?? logFactory.ThrowExceptions);
+            }
+
+            public ValidatedConfigurationElement(ILoggingConfigurationElement element, bool throwConfigExceptions)
+            {
+                _throwConfigExceptions = throwConfigExceptions;
+                Name = element.Name.Trim();
+                ValueLookup = CreateValueLookup(element, throwConfigExceptions);
+                _element = element;
+            }
+
+            public string Name { get; }
+
+            public IDictionary<string, string> ValueLookup { get; }
+
+            public IEnumerable<ValidatedConfigurationElement> ValidChildren
+            {
+                get
+                {
+                    if (_validChildren != null)
+                        return _validChildren;
+                    else
+                        return YieldValidChildren();
+                }
+            }
+
+            IEnumerable<ValidatedConfigurationElement> YieldValidChildren()
+            {
+                foreach (var child in _element.Children)
+                {
+                    _validChildren = _validChildren ?? new List<ValidatedConfigurationElement>();
+                    var validChild = new ValidatedConfigurationElement(child, _throwConfigExceptions);
+                    _validChildren.Add(validChild);
+                    yield return validChild;
+                }
+                _validChildren = _validChildren ?? ArrayHelper.Empty<ValidatedConfigurationElement>();
```

Comment:
This is a bit unexpected (according to the method name)

could we move this out of this method, or rename the method (YieldAndCacheValidChildren or something like that)

Reply:
Will rename the method. But the reason is to perform lazy validation of children.